spec/design_topics/data_interchange.md

```diff
@@ -58,18 +58,67 @@ means the object it is attached to must return a `numpy.ndarray`
+containing `from_dlpack` (the "consumer"). If possible, this must be
+zero-copy (i.e. `y` will be a _view_ on `x`). If not possible, that library
+may make a copy of the data. In both cases:
+- the producer keeps owning the memory
+- `y` may or may not be a view, therefore the user must keep the
+   recommendation to avoid mutating `y` in mind - see
+   {ref}`copyview-mutability`.
+- Both `x` and `y` may continue to be used just like arrays created in other ways.
 
 If an array that is accessed via the interchange protocol lives on a
-device that the requesting library does not support, one of two things
-must happen: moving data to another device, or raising an exception.
-Device transfers are typically expensive, hence doing that silently can
-lead to hard to detect performance issues. Hence it is recommended to
-raise an exception, and let the user explicitly enable device transfers
-via, e.g., a `force=False` keyword that they can set to `True`.
-```
+device that the requesting library does not support, it is recommended to
+raise a `TypeError`.
+
+
+### Implementation
+
+_Note that while this API standard largely tries to avoid discussing implementation details, some discussion and requirements are needed here because data interchange requires coordination between implementers on, e.g., memory management._
+
+![Diagram of DLPack structs](/_static/images/DLPack_diagram.png)
+
+_DLPack diagram. Dark blue are the structs it defines, light blue struct members, gray text enum values of supported devices and data types._
+
+The `__dlpack__` method will produce a `PyCapsule` containing a
+`DLPackManagedTensor`, which will be consumed immediately within
+`from_dlpack` - therefore it is consumed exactly once, and it will not be
+visible to users of the Python API.
+
+The consumer must set the PyCapsule name to `"used_dltensor"`, and call the
+`deleter` of the `DLPackManagedTensor` when it no longer needs the data.
+
+For further details on DLPack design and how to implement support for it,
+refer to [github.com/dmlc/dlpack](https://github.com/dmlc/dlpack).
+
+:::{warning}
+DLPack contains a `device_id`, which will be the device ID (an integer, `0, 1, ...`) which the producer library uses. In practice this will likely be the same numbering as that of the consumer, however that is not guaranteed. Depending on the hardware type, it may be possible for the consumer library implementation to look up the actual device from the pointer to the data - this is possible for example for CUDA device pointers.
+
+It is recommended that implementers of this array API document whether the
+`.device` attribute of the array returned from `from_dlpack` is guaranteed to
+be a certain order or not.
+:::
+
+## Out of scope for data interchange
+
+- Stream handling, e.g. for CUDA streams. TBD: if DLPack gains
+  synchronization semantics support (see
+  [dlpack/issues/57](https://github.com/dmlc/dlpack/issues/57)), details on
```

Comment:
I think we could add a description for the current synchronization semantics:

DLPack implicitly requires both the producer and consumer follows the same CUDA stream. An explicit synchronization is required by consumer or the producer if a different stream is used.

Reply:
Yes, good idea. 

It actually took me a little while to fully parse the discussion in https://github.com/dmlc/dlpack/issues/57, so let me see if I got this right:
- If either the producer or the consumer uses the legacy default stream, or if they're both on the same stream, there is no issue.
- If both producer and consumer are on non-default streams, and they're not on the same stream, there may be a silent correctness issue.
- The producer and consumer cannot know of each other what stream they use.

In addition, I think most users of the Python API won't know either, and the libraries they use may not expose an API to do a manual sync.

> An explicit synchronization is required by consumer or the producer if a different stream is used.

So, to avoid correctness issues, one of the producer or consumer must _always_ synchronize. Making the producer do it seems to make most sense (it's conceptually similar to, for example, materializing data to CPU memory in a library that does delayed evaluation). Hence should we add this as in implementation requirement to `__dlpack__`?

Reply:
I'm sure not everyone will like that, so maybe there's a way to make sure such an implementation can be upgraded to a future more complex method that knows about streams? Maybe that's a reason to make the consumer synchronize - it can then avoid synchronizing if a stream number appears?

Reply:
Here is a strawman proposal (will cross post to the dlpack discussion thread) per S0 style in 
```
def __dlpack__(stream=None):
    """Get a DLTensor capsule that is visible in stream.
   
    The producer will take charge to do dependency synchronize to stream.
    If no stream is None, then it defaults to the default stream(0 in cuda).
    """
    pass
```

Reply:
Updated, so resolving this comment. I'll keep tracking https://github.com/dmlc/dlpack/issues/57 if anything changes there.

Reply:
> I'm sure not everyone will like that, so maybe there's a way to make sure such an implementation can be upgraded to a future more complex method that knows about streams? Maybe that's a reason to make the consumer synchronize - it can then avoid synchronizing if a stream number appears?

Congrats, @rgommers! You're on the thinking process of arriving at a CAI-v3-like spec 🙂

> I would recommend to go with S0 style API

I will follow up this issue in dmlc/dlpack#57 too, but just for the record: @tqchen, S0 does not solve the problem if a Producer does not expose its internal streams, see my objections in https://github.com/dmlc/dlpack/issues/57#issuecomment-752185385. This was one of the main reasons that CAI v3 requires Producer to provide a semantically correct handle (a stream pointer) for Consumer to sync over.